zio-http/src/main/scala/zio/http/Body.scala

```diff
@@ -43,58 +45,83 @@ trait Body { self =>
 
   def isComplete: Boolean
 
-  def write(ctx: Ctx)(implicit trace: Trace): Task[Boolean]
 }
 
 object Body {
 
+  private[zio] sealed trait UnsafeWriteable extends Body
+
+  private[zio] sealed trait UnsafeBytes extends Body {
+    private[zio] def unsafeAsArray(implicit unsafe: Unsafe): Array[Byte]
+  }
+
   /**
    * Helper to create empty Body
    */
-  val empty: Body = new Body {
-    override def asChunk(implicit trace: Trace): Task[Chunk[Byte]]              = ZIO.succeed(Chunk.empty[Byte])
+  val empty: Body = EmptyBody
+
+  private[zio] object EmptyBody extends Body with UnsafeWriteable with UnsafeBytes {
+
+    override def asArray(implicit trace: Trace): Task[Array[Byte]] = ZIO.succeed(Array.empty)
+
+    override def asChunk(implicit trace: Trace): Task[Chunk[Byte]] = ZIO.succeed(Chunk.empty[Byte])
+
     override def asStream(implicit trace: Trace): ZStream[Any, Throwable, Byte] = ZStream.empty
-    override def write(ctx: Ctx)(implicit trace: Trace): Task[Boolean]          = ZIO.succeed(false)
     override def isComplete: Boolean                                            = true
 
     override def toString(): String = "Body.empty"
+
+    override private[zio] def unsafeAsArray(implicit unsafe: Unsafe): Array[Byte] = Array.empty[Byte]
   }
 
   /**
    * Helper to create Body from AsciiString
    */
-  def fromAsciiString(asciiString: AsciiString): Body = new Body {
-    override def isComplete: Boolean = true
+  def fromAsciiString(asciiString: AsciiString): Body = AsciiStringBody(asciiString)
```

Comment:
`AsciiString` is a netty thing, that exists for smaller / faster encoding / decoding. It would be a shame for such a Netty detail to leak into `zio.http.Body`. 

I think the best we should probably do is `CharSequenceBody`. Then on the Netty backend we could type case to see if it's an `AsciiString`. Then make some effort to actually use `AsciiString` as the `CharSequence` where useful to improving performance.